### ISSUE TITLE
Support URLs that identify created object at end of URL

### ISSUE THREAD

**Issue by rberlind** [Mastercard/terraform-provider-restapi#5]
Some APIs such as the Vault API require the user to specify the ID of an object being created as the last part of the URL.  While this provider can actually create a Vault secret, it gets an error when trying to read the secret back during `terraform apply`.  So, if I set id_attribute to "id", my resource path to "/secret/restapi/test" and data to "{\"foo\":\"bar\",\"id\":\"test\"}", the provider successfully writes the secret to <vault_url>:8200/v1/secret/restapi/test, but then tries to read it back from <vault_url>:8200/v1/secret/restapi/test/test which does not exist.

Additionally, one would not normally include the final part of the Vault path as a key in the body as I have done.  I only did that so I could get the provider to write the secret.  Normally, my data would have just been "{\"foo\":\"bar\"}"

So, I think it would be useful to somehow support URLs like these.  Note that these URLs each effectively only manage a single object.

The key to making the provider work with these paths is to make it use the same URL for POSTs that it uses for GETs, PUTs, and DELETEs.  Somehow, the user would be able to indicate that the ID for their object actually is the last segment of the resource path.

See https://github.com/Mastercard/terraform-provider-restapi/issues/4 in which I requested more flexibility for specifying object IDs and included the possibility of specifying that the ID is the last segment of the resource path.

**Comment by DRuggeri** [comment]
I think it would definitely be possible to allow for overriding GET/PUT/DELETE paths in the resource object. This could be accomplished trivially and would just use `path` as a default, but permit something more specific.

I'm thinking `create_path`, `destroy_path` and `read_path` would be fine attributes for this. Thoughts?

**Comment by rberlind** [comment]
Those would certainly work.  Thanks.

**Comment by rberlind** [comment]
Would you also need an update_path?

**Comment by DRuggeri** [comment]
Yep - sorry about that. It would be all four of the CRUD operations.

**Comment by rberlind** [comment]
Thanks so much for adding this.  Really appreciate it.  I will test tomorrow with the Vault API, but more interesting test will come next week when the customer I'm working with tests with vRO.

In api_object.go, if the specific paths are used, then I don't think you should add obj.id.  Instead, you should assume that the path already includes the object ID.

Should lines 62-65 of api_object_test.go be more specific?  I know it is just for testing.

**Comment by rberlind** [comment]
What you have added here will work well for an API like Vault in which the initial URL gives the ID of the object being created IF you do avoid adding the ID to the new paths.  I actually think you could probably have just added a single path attribute like "modify_path" instead of the 4 you added since I suspect that the get, put and delete paths will always be the same.  On the other hand, it is hard to think of a correct name for a combined "get,put, delete" path.  While I just used "modify_path", that is not quite right since read does not modify the object. "post_creation_path" might work, but sounds strange. Also, it is possible that some APIs actually would need different paths for get, put, and delete.  So, the approach you took is probably best.

**Comment by rberlind** [comment]
Looking at your code some more, I see that simply suppressing the ID from the paths your do GET, PUT, and DELETE operations against is not so easy given the way you currently set the paths.  I think the easiest way of doing this would be to actually allow obj.id to be "".  I realize, however, that you currently have multiple checks to avoid that.  Perhaps the solution would be to add one more boolean attribute to the resource called "null_id_allowed" and then check for that being set to true in your checks that currently check for obj.id == "". Alternatively, you could allow the special value of "NULL" in the id_attribute attribute and check for that.

To be honest, I think it is desirable to allow null IDs in any case.  It would particularly be useful for an API like Vault in which the ID is the last part of the URL.  As I mentioned in the first post on this issue, I currently am providing the ID in the data object in a way that is not natural for Vault. If I can both specify the read_path and indicate that the ID should be empty or null, then I could avoid that.

**Comment by DRuggeri** [comment]
I think the vault use case is worth pursuing. I've made a modification in the latest release to allow specifying if/where the ID gets injected into the `*_path` attributes. I've also made it so you can manually specify the `id` attribute.

Unfortunately, there's no way to have a null or empty ID - the IDs are integral to terraform's data structures.

**Comment by rberlind** [comment]
Thanks so much for making the changes to support the Vault usecase.  Apologies for not getting back to you sooner.  I did successfully test today with the code below which allowed me to create, update, and delete (destroy) a secret in Vault.

```
resource "restapi_object" "tfe_vault_secret" {
  path = "/secret/restapi"
  create_path = "/secret/restapi/{id}"
  read_path = "/secret/restapi/{id}"
  update_path = "/secret/restapi/{id}"
  destroy_path = "/secret/restapi/{id}"
  data = "{\"foo\":\"bar-updated\",\"id\":\"test\"}"
  debug = true
}
```

I also then tried using the new object_id attribute of the resource which I believe is a new attribute intended to let the user set the object id manually, but was less successful.  I got an error saying "Provided data does not have id attribute for the object's id and the client is not configured to read the object from a POST response. Without an id, the object cannot be managed."

I think the problem is on line 40 of common.go where you have "d.ID()" instead of "id".

I will try to do a pull request to fix this myself.

Thanks again for all your responsiveness.
Roger

**Comment by rberlind** [comment]
OK.  I submitted #10 for this.

I had some issues working with the code after forking it because main.go has the import "github.com/Mastercard/terraform-provider-restapi/restapi".  So, after doing the Go command `go get github.com\Mastercard\terraform-provider-restapi` and making changes in my own fork under "github.com\rberlind\terraform-provider-restapi", the changes were not applied.

Eventually, I figured out that one has to follow one of several processes to work around this issue.  In particular, following https://medium.com/@sasom/how-to-fork-go-packages-f863e370bbb6, I set an additional remote with `git remote add fork https://github.com/rberlind/terraform-provider-restapi.git` under /Users/roger/goworkspace/src/github.com/Mastercard/terraform-provider-restapi and then ran `git pull fork master` to pull my pushed changes into this directory.  That way, when I built the provider, it used my new code.

A downside of this approach is that I had to push my changes to my fork and then pull them from there into /Users/roger/goworkspace/src/github.com/Mastercard/terraform-provider-restapi each time I made a change. I also found https://dev.to/loderunner/working-with-forks-in-go-3ab6 which suggested adding the command `git branch -u fork/master` so that I could work directly under that directory and avoid the need to push and pull for each change.  Of course, I then had to be careful to use `git push fork master` instead of `git push origin master`.

Ideally, it would be nice to avoid the line `"github.com/Mastercard/terraform-provider-restapi/restapi"` in main.go.  I tried to find a way of providing a relative path that would point to restapi, but that does not appear possible in Go. 

I also explored using [dep](https://github.com/golang/dep) which allows you to specify a different source for any package, but backed away from that after having some problems.  But I think those problems were caused by my temporarily running `go build` from the restapi directory instead of the one above it.  So, I might try using dep again.

**Comment by rberlind** [comment]
I think we could have closed this a long time ago since you added the various paths. And I closed #10 quite some time ago.  I would still like to see the master fork/branch of this code support what I have implemented in my fork for getting the object ID from a specific header, including a header in which the ID is the last segment of a URL contained in the header.  But I know that supporting it has ramifications since it requires passing both headers and body back from calls to https://github.com/Mastercard/terraform-provider-restapi/blob/master/restapi/api_client.go#L106
